### PR TITLE
feat: add circuit sectors

### PR DESCRIPTION
**Pull request by tooboredtocode**
This PR adds a coloured circuit/sectors.

The way the track is split into parts is a bit ugly though, when have time I might create custom components for the tracks with proper sectors, marshal sectors and the pitlane.

### CODE REVIEW

**Review by slowlydev — CHANGES_REQUESTED** [review]
Thanks a lot for this cool PR, I appreciate it a lot.
I added a few things, just some minor things...

**Comment by slowlydev** [comment]
I also saw on your github profile that u code a lot of rust, If u want to I would appreciate if you could take a look at the [feature/v2 ](https://github.com/slowlydev/f1-dash/tree/feature/v2) branch where I rewrote the backend in rust, Im sure there is some room for improvement for my rust code.

File: dash/src/components/Map.tsx [hunk 1]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+		return "#44DDEE";
+	}
+	return "#FFFFFF";
+}
+
+const createSectors = (data: MapType) => {
+	const sectors: Sector[] = [];
+	const points: TrackPosition[] = data.x.map((x, index) => ({ x, y: data.y[index] }));
+
+	for (let i = 0; i < data.marshalSectors.length; i++) {
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:94]: We're looping, so no...

File: dash/src/components/Map.tsx [hunk 2]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+		sectors.push({
+			number: i + 1,
+			start: data.marshalSectors[i].trackPosition,
+			end: data.marshalSectors[i + 1] ? data.marshalSectors[i + 1].trackPosition : data.marshalSectors[0].trackPosition,
+			points: []
+		});
+	}
+
+	let dividers: number[] = sectors.map(s => findMinDistance(s.start, points));
+	for (let i = 0; i < dividers.length; i++) {
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:104]: We're looping, so no...

File: dash/src/components/Map.tsx [hunk 3]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+	if (msg.flag === "RED") {
+		return "#930013";
+	}
+	if (msg.category === "SafetyCar") {
+		return "#44DDEE";
+	}
+	return "#FFFFFF";
+}
+
+const createSectors = (data: MapType) => {
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:90]: done

File: dash/src/components/Map.tsx [hunk 4]
```diff
@@ -91,13 +188,24 @@ export default function Map({ circuitKey, positionBatches }: Props) {
...
 			/>
 
-			<path
-				stroke="white"
-				strokeWidth={60}
-				strokeLinejoin="round"
-				fill="transparent"
-				d={`M${points[0].x},${points[0].y} ${points.map((point) => `L${point.x},${point.y}`).join(" ")}`}
-			/>
+			{<>{
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:191]: done

File: dash/src/components/Map.tsx [hunk 5]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+		return msg.category === "SafetyCar";
+	})
+
+	if (!msg || msg.flag === "CLEAR") {
+		return "#FFFFFF";
+	}
+	if (msg.flag === "YELLOW" || msg.flag === "DOUBLE YELLOW") {
+		return "#ffff14";
+	}
+	// @ts-ignore
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:80]: my bad, that was before I added "RED" to the avaliable flag colours

File: dash/src/components/Map.tsx [hunk 6]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+		const distance = calculateDistance(point.x, point.y, points[i].x, points[i].y);
+		if (distance < min) {
+			min = distance;
+			minIndex = i;
+		}
+	}
+	return minIndex;
+}
+
+const sector_color = (sector: number, messages: RaceControlMessageType[] | undefined) => {
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx:63]: done

File: dash/src/components/Map.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (slowlydev) [reply dash/src/components/Map.tsx:94]: Ah right my bad, I am too used to "for...in" and "for...of" loops... .-.

**Comment by tooboredtocode** [comment]
@slowlydev im trying to test some things, do you have any data I can use with the simulator, preferably with all possible situations (red flag, VSC, yellow flag, etc.)?

**Comment by slowlydev** [comment]
> @slowlydev im trying to test some things, do you have any data I can use with the simulator, preferably with all possible situations (red flag, VSC, yellow flag, etc.)?

I don't have a synthetic file for all possible situations. But here are some files form data-saver.

https://github.com/slowlydev/f1-dash-data-parser/releases/tag/data

I also got some data from Japan but only started "recording" during the redflag. And I noticed that the sectors where the crash happened (and where yellow flags were once) were still white when the rest of the track is red.

File: dash/src/components/Map.tsx [hunk 7]
```diff
@@ -32,10 +36,90 @@ const rotate = (x: number, y: number, a: number, px: number, py: number) => {
...
+			return ["CLEAR", "YELLOW", "DOUBLE YELLOW", "RED"].includes(msg.flag);
+		}
+		return msg.category === "SafetyCar";
+	})
+
+	if (!msg || msg.flag === "CLEAR") {
+		return "#FFFFFF";
+	}
+	if (msg.flag === "YELLOW" || msg.flag === "DOUBLE YELLOW") {
+		return "#ffff14";
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/components/Map.tsx]: switched to tailwind colors

**Comment by tooboredtocode** [comment]
Fixed the issues by syncing the track color to the trackState and only falling back to the raceControlMessages to figure out which sectors are yellow

**Review by slowlydev — CHANGES_REQUESTED** [review]
Thanks for the adjustments :)
Added some more comments, I like that we are using the track status now, feels a lot cleaner this way.

File: dash/src/lib/getTrackStatusMessage.ts [hunk 8]
```diff
@@ -1,6 +1,9 @@
 type StatusMessage = {
 	message: string;
 	color: string;
+	trackColor: string;
+	bySector?: boolean;
+	pulse?: number;
```
[the remark this answers is not in the document]
Comment (tooboredtocode) [reply dash/src/lib/getTrackStatusMessage.ts:6]: I was testing, adding it for the various safety cars, however I wasn't really satisfied with it, nor sure if ppl would like it, so I removed the setting. However I thought that deleting the code would be a shame, perhaps it is something that could be interesting...

**Comment by tooboredtocode** [comment]
Cleaned the code up a bit, we now use an effect to calculate everything. Which is definitely much cleaner

File: dash/src/lib/getTrackStatusMessage.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (slowlydev) [reply dash/src/lib/getTrackStatusMessage.ts:6]: Okay yeah I also find it a interesting idea, I guess we can leave it for now.

**Review by slowlydev — APPROVED** [review]
Sorry thought we had pending requested changes, but just went through the changes again and it looks all good. 🎉

**Pull request merged**